mock/service/controller.go

```diff
@@ -338,21 +338,183 @@ func (s *service) ControllerGetCapabilities(
+		return nil, status.Error(codes.InvalidArgument, "Name missing in request")
+	}
+	if len(req.GetSourceVolumeId()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "SourceVolumeId missing in request")
+	}
+
+	// Check to see if the snapshot already exists.
+	if i, v := s.findSnapshotByName(ctx, req.Name); i >= 0 {
+		// Requested snapshot name already exists
+		if v.GetSourceVolumeId() == req.SourceVolumeId {
+			return nil, status.Error(codes.AlreadyExists,
+				fmt.Sprintf("Snapshot with name %s already exists", req.GetName()))
+		}
+		return &csi.CreateSnapshotResponse{Snapshot: &v}, nil
+	}
+
+	// Create the snapshot and add it to the service's in-mem snapshot slice.
+	snapshot := s.newSnapshot(req.Name, req.SourceVolumeId)
+	s.snapshotsRWL.Lock()
+	defer s.snapshotsRWL.Unlock()
+	s.snapshots = append(s.snapshots, snapshot)
+
+	return &csi.CreateSnapshotResponse{Snapshot: &snapshot.SnapshotCSI}, nil
 }
 
 func (s *service) DeleteSnapshot(ctx context.Context,
 	req *csi.DeleteSnapshotRequest) (*csi.DeleteSnapshotResponse, error) {
-	return nil, status.Error(codes.InvalidArgument, "Not Implemented")
+
+	s.snapshotsRWL.Lock()
+	defer s.snapshotsRWL.Unlock()
+
+	//  If the snapshot is not specified, return error
+	if len(req.SnapshotId) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Snapshot ID cannot be empty")
+	}
+
+	// If the snapshot does not exist then return an idempotent response.
+	i, _ := s.findSnapshotNoLock("id", req.SnapshotId)
+	if i < 0 {
+		return &csi.DeleteSnapshotResponse{}, nil
+	}
+
+	// This delete logic preserves order and prevents potential memory
+	// leaks. The slice's elements may not be pointers, but the structs
+	// themselves have fields that are.
+	copy(s.snapshots[i:], s.snapshots[i+1:])
+	s.snapshots[len(s.snapshots)-1] = Snapshot{}
+	s.snapshots = s.snapshots[:len(s.snapshots)-1]
+	log.WithField("SnapshotId", req.SnapshotId).Debug("mock delete snapshot")
+	return &csi.DeleteSnapshotResponse{}, nil
 }
 
 func (s *service) ListSnapshots(ctx context.Context,
 	req *csi.ListSnapshotsRequest) (*csi.ListSnapshotsResponse, error) {
-	return nil, status.Error(codes.InvalidArgument, "Not Implemented")
+
+	// case 1: SnapshotId is not empty, return snapshots that match the snapshot id.
+	if len(req.GetSnapshotId()) != 0 {
+		i, snapshot := s.findSnapshotNoLock("id", req.SnapshotId)
+		if i < 0 {
+			return &csi.ListSnapshotsResponse{}, nil
+		}
+		return &csi.ListSnapshotsResponse{
+			Entries: []*csi.ListSnapshotsResponse_Entry{
+				{
+					Snapshot: &snapshot,
+				},
+			},
+		}, nil
+	}
+
+	// case 2: SourceVolumeId is not empty, return snapshots that match the source volume id.
+	if len(req.SourceVolumeId) != 0 {
+		i, snapshot := s.findSnapshotNoLock("id", req.SourceVolumeId)
+		if i < 0 {
+			return &csi.ListSnapshotsResponse{}, nil
+		}
+		return &csi.ListSnapshotsResponse{
+			Entries: []*csi.ListSnapshotsResponse_Entry{
+				{
+					Snapshot: &snapshot,
+				},
+			},
+		}, nil
+	}
```

Comment:
I think we need to consider both `req.GetSnapshotId()` and `req.GetSourceVolumeID` when they aren't empty. If the volume received from ID has a different source volume ID, do not include them in the snapshot list.